### PR TITLE
[exporter/clickhouse] Add integration test

### LINKED ISSUE #29626: [exporter/clickhouse] Can't create metrics table with the new `ttl` configuration.
### Component(s)

exporter/clickhouse

## Description

With `0.90.0` collector, when starting with clickhouse exporter, collector can't create metrics tables in clickhouse. It works well with `0.89.0` version, so this bug should be introduced by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29095.

## Steps to Reproduce

- Start a clean clickhouse server locally.
- `docker run --rm -v \`pwd\`/otel.yaml:/etc/otelcol-contrib/config.yaml --network host ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib:0.90.0`

## Expected Result

The collector image should work well with the configuration.

## Actual Result

```
collector server run finished with error: cannot start pipelines: exec create metrics table sql: code: 47, message: Missing columns: 'Timestamp' while processing query: 'toDateTime(Timestamp) + toIntervalDay(3)', required columns: 'Timestamp' 'Timestamp'
```

### Collector version

v0.90.0

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

```yaml
receivers:
  otlp:
    protocols:
      grpc:
      http:

processors:
  batch:
    timeout: 5s
    send_batch_size: 100000

exporters:
  clickhouse:
    endpoint: clickhouse://localhost:9000?dial_timeout=10s&compress=lz4
    username: default
    password: ""
    database: otel
    ttl: 72h
    logs_table_name: otel_logs
    traces_table_name: otel_traces
    metrics_table_name: otel_metrics
    timeout: 5s
    retry_on_failure:
      enabled: true
      initial_interval: 5s
      max_interval: 30s
      max_elapsed_time: 300s

extensions:
  health_check:
  pprof:
  zpages:

service:
  extensions: [health_check, pprof, zpages]
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [clickhouse]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [clickhouse]
    logs:
      receivers: [otlp]
      processors: [batch]
      exporters: [clickhouse]
```

### Log output

```shell
2023-12-02T16:15:14.022Z	info	service@v0.90.0/telemetry.go:86	Setting up own telemetry...
2023-12-02T16:15:14.026Z	info	service@v0.90.0/telemetry.go:203	Serving Prometheus metrics	{"address": ":8888", "level": "Basic"}
2023-12-02T16:15:14.033Z	info	service@v0.90.0/service.go:148	Starting otelcol-contrib...	{"Version": "0.90.0", "NumCPU": 2}
2023-12-02T16:15:14.034Z	info	extensions/extensions.go:34	Starting extensions...
2023-12-02T16:15:14.034Z	info	extensions/extensions.go:37	Extension is starting...	{"kind": "extension", "name": "zpages"}
2023-12-02T16:15:14.034Z	info	zpagesextension@v0.90.0/zpagesextension.go:53	Registered zPages span processor on tracer provider	{"kind": "extension", "name": "zpages"}
2023-12-02T16:15:14.034Z	info	zpagesextension@v0.90.0/zpagesextension.go:63	Registered Host's zPages	{"kind": "extension", "name": "zpages"}
2023-12-02T16:15:14.037Z	info	zpagesextension@v0.90.0/zpagesextension.go:75	Starting zPages extension	{"kind": "extension", "name": "zpages", "config": {"TCPAddr":{"Endpoint":"localhost:55679"}}}
2023-12-02T16:15:14.037Z	info	extensions/extensions.go:45	Extension started.	{"kind": "extension", "name": "zpages"}
2023-12-02T16:15:14.037Z	info	extensions/extensions.go:37	Extension is starting...	{"kind": "extension", "name": "pprof"}
2023-12-02T16:15:14.043Z	info	pprofextension@v0.90.0/pprofextension.go:60	Starting net/http/pprof server	{"kind": "extension", "name": "pprof", "config": {"TCPAddr":{"Endpoint":"localhost:1777"},"BlockProfileFraction":0,"MutexProfileFraction":0,"SaveToFile":""}}
2023-12-02T16:15:14.043Z	info	extensions/extensions.go:45	Extension started.	{"kind": "extension", "name": "pprof"}
2023-12-02T16:15:14.047Z	info	extensions/extensions.go:37	Extension is starting...	{"kind": "extension", "name": "health_check"}
2023-12-02T16:15:14.047Z	info	healthcheckextension@v0.90.0/healthcheckextension.go:35	Starting health_check extension	{"kind": "extension", "name": "health_check", "config": {"Endpoint":"0.0.0.0:13133","TLSSetting":null,"CORS":null,"Auth":null,"MaxRequestBodySize":0,"IncludeMetadata":false,"ResponseHeaders":null,"Path":"/","ResponseBody":null,"CheckCollectorPipeline":{"Enabled":false,"Interval":"5m","ExporterFailureThreshold":5}}}
2023-12-02T16:15:14.047Z	warn	internal@v0.90.0/warning.go:40	Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks	{"kind": "extension", "name": "health_check", "documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
2023-12-02T16:15:14.049Z	info	extensions/extensions.go:45	Extension started.	{"kind": "extension", "name": "health_check"}
2023-12-02T16:15:14.125Z	info	service@v0.90.0/service.go:188	Starting shutdown...
2023-12-02T16:15:14.125Z	info	healthcheck/handler.go:132	Health Check state change	{"kind": "extension", "name": "health_check", "status": "unavailable"}
2023-12-02T16:15:14.125Z	info	extensions/extensions.go:52	Stopping extensions...
2023-12-02T16:15:14.125Z	info	zpagesextension@v0.90.0/zpagesextension.go:98	Unregistered zPages span processor on tracer provider	{"kind": "extension", "name": "zpages"}
2023-12-02T16:15:14.125Z	info	service@v0.90.0/service.go:202	Shutdown complete.
Error: cannot start pipelines: exec create metrics table sql: code: 47, message: Missing columns: 'Timestamp' while processing query: 'toDateTime(Timestamp) + toIntervalDay(3)', required columns: 'Timestamp' 'Timestamp'
2023/12/02 16:15:14 collector server run finished with error: cannot start pipelines: exec create metrics table sql: code: 47, message: Missing columns: 'Timestamp' while processing query: 'toDateTime(Timestamp) + toIntervalDay(3)', required columns: 'Timestamp' 'Timestamp'
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Frapschen**
**Description:** <Describe what has changed.>

as https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29626#issuecomment-1839884566 mentioned, this PR adds an integration test for clickhouse exporter and it brings those benefits:
- Verify tables are properly created in clickhouse after any breaking change(filter out bug like [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29573)  before merge)
- Verify if any bugs occur during data insertion like missed data or insert into an error column([30210](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30210)) .etc.
- Compatibility testing for `clickhouse-server:23-alpine` and `clickhouse-server:22-alpine`

close https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29626

### CODE REVIEW

**Comment by Frapschen** [comment]
wait for https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30210 merge

**Comment by Frapschen** [comment]
The test result in my local env:
```
=== RUN   TestIntegration
--- PASS: TestIntegration (112.81s)
=== RUN   TestIntegration/test_clickhouse_23-alpine
2024/01/19 10:33:48 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 23.0.5
  API Version: 1.42
  Operating System: Docker Desktop
  Total Memory: 7850 MB
  Resolved Docker Host: unix:///Users/fraps/.docker/run/docker.sock
  Resolved Docker Socket Path: /var/run/docker.sock
  Test SessionID: cebe223d56bab5923ac8f3f77d1afaca0ef1a5890da5b761ed98737310863834
  Test ProcessID: f4efcbf0-930a-44e1-aec9-ccc9842b935b
2024/01/19 10:33:57 Failed to pull image: Error response from daemon: Get "https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/sha256:739f19eb83b1710f22ff452d7bec9cdd3f53d0ba5ff4096c6516a85de88eba0c": EOF, will retry
2024/01/19 10:34:02 Failed to pull image: Error response from daemon: Get "https://registry-1.docker.io/v2/": EOF, will retry
2024/01/19 10:34:11 Failed to pull image: Error response from daemon: Head "https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/0.6.0": Get "https://auth.docker.io/token?account=frapschen&scope=repository%3Atestcontainers%2Fryuk%3Apull&service=registry.docker.io": EOF, will retry
2024/01/19 10:34:18 Failed to pull image: Error response from daemon: Get "https://registry-1.docker.io/v2/": EOF, will retry
2024/01/19 10:34:33 Failed to pull image: Error response from daemon: Head "https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/0.6.0": EOF, will retry
2024/01/19 10:34:42 Failed to pull image: Error response from daemon: Head "https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/0.6.0": Get "https://auth.docker.io/token?account=frapschen&scope=repository%3Atestcontainers%2Fryuk%3Apull&service=registry.docker.io": EOF, will retry
2024/01/19 10:34:51 Failed to pull image: Error response from daemon: Head "https://registry-1.docker.io/v2/testcontainers/ryuk/manifests/0.6.0": Get "https://auth.docker.io/token?account=frapschen&scope=repository%3Atestcontainers%2Fryuk%3Apull&service=registry.docker.io": EOF, will retry
2024/01/19 10:35:02 Failed to pull image: Error response from daemon: Get "https://registry-1.docker.io/v2/": EOF, will retry
2024/01/19 10:35:30 🐳 Creating container for image testcontainers/ryuk:0.6.0
2024/01/19 10:35:30 ✅ Container created: 1650ba8a9c94
2024/01/19 10:35:30 🐳 Starting container: 1650ba8a9c94
2024/01/19 10:35:32 ✅ Container started: 1650ba8a9c94
2024/01/19 10:35:32 🚧 Waiting for container id 1650ba8a9c94 image: testcontainers/ryuk:0.6.0. Waiting for: &{Port:8080/tcp timeout:<nil> PollInterval:100ms}
2024/01/19 10:35:32 🐳 Creating container for image clickhouse/clickhouse-server:23-alpine
2024/01/19 10:35:32 ✅ Container created: bcc223138ff2
2024/01/19 10:35:32 🐳 Starting container: bcc223138ff2
2024/01/19 10:35:33 ✅ Container started: bcc223138ff2
2024/01/19 10:35:33 🚧 Waiting for container id bcc223138ff2 image: clickhouse/clickhouse-server:23-alpine. Waiting for: &{Port:9000 timeout:0x1400061b390 PollInterval:100ms}
2024/01/19 10:35:36 🐳 Starting container: bcc223138ff2
2024/01/19 10:35:36 ✅ Container started: bcc223138ff2
2024/01/19 10:35:36 🚧 Waiting for container id bcc223138ff2 image: clickhouse/clickhouse-server:23-alpine. Waiting for: &{Port:9000 timeout:0x1400061b390 PollInterval:100ms}
    logger.go:130: 2024-01-19T10:35:37.033+0800	DEBUG	insert logs	{"records": 1, "cost": "90.344042ms"}
    logger.go:130: 2024-01-19T10:35:37.162+0800	DEBUG	insert traces	{"records": 1, "cost": "88.239416ms"}
    logger.go:130: 2024-01-19T10:35:37.257+0800	DEBUG	insert sum metrics	{"records": 1, "cost": "38.429292ms"}
    logger.go:130: 2024-01-19T10:35:37.268+0800	DEBUG	insert exponential histogram metrics	{"records": 1, "cost": "49.040958ms"}
    logger.go:130: 2024-01-19T10:35:37.268+0800	DEBUG	insert histogram metrics	{"records": 1, "cost": "49.128167ms"}
    logger.go:130: 2024-01-19T10:35:37.292+0800	DEBUG	insert summary metrics	{"records": 1, "cost": "73.559958ms"}
    --- PASS: TestIntegration/test_clickhouse_23-alpine (109.09s)
=== RUN   TestIntegration/test_clickhouse_22-alpine
2024/01/19 10:35:37 🐳 Creating container for image clickhouse/clickhouse-server:22-alpine
2024/01/19 10:35:37 ✅ Container created: a734a0325846
2024/01/19 10:35:37 🐳 Starting container: a734a0325846
2024/01/19 10:35:38 ✅ Container started: a734a0325846
2024/01/19 10:35:38 🚧 Waiting for container id a734a0325846 image: clickhouse/clickhouse-server:22-alpine. Waiting for: &{Port:9000 timeout:0x140003372d0 PollInterval:100ms}
2024/01/19 10:35:40 🐳 Starting container: a734a0325846
2024/01/19 10:35:40 ✅ Container started: a734a0325846
2024/01/19 10:35:40 🚧 Waiting for container id a734a0325846 image: clickhouse/clickhouse-server:22-alpine. Waiting for: &{Port:9000 timeout:0x140003372d0 PollInterval:100ms}
    logger.go:130: 2024-01-19T10:35:40.669+0800	DEBUG	insert logs	{"records": 1, "cost": "102.156875ms"}
    logger.go:130: 2024-01-19T10:35:40.860+0800	DEBUG	insert traces	{"records": 1, "cost": "82.321333ms"}
    logger.go:130: 2024-01-19T10:35:41.012+0800	DEBUG	insert histogram metrics	{"records": 1, "cost": "53.820667ms"}
    logger.go:130: 2024-01-19T10:35:41.012+0800	DEBUG	insert summary metrics	{"records": 1, "cost": "53.859125ms"}
    logger.go:130: 2024-01-19T10:35:41.015+0800	DEBUG	insert sum metrics	{"records": 1, "cost": "56.71675ms"}
    logger.go:130: 2024-01-19T10:35:41.026+0800	DEBUG	insert exponential histogram metrics	{"records": 1, "cost": "66.981208ms"}
    --- PASS: TestIntegration/test_clickhouse_22-alpine (3.73s)
PASS
```

**Review by hanjm — APPROVED** [review]
LGTM

**Comment by Frapschen** [comment]
why there are so many go mod files changed after I `make gotidy`?

**Comment by dmitryax** [comment]
PTAL at the failing CI. looks like your change flips `cloud.google.com/go` dependancy

**Comment by Frapschen** [comment]
@dmitryax please review

File: .chloggen/add_integration_test_for_clickhous_exporter.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: clickhouseexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add ingetration test for clickhouse exporter
```
Comment (codeboten) [inline .chloggen/add_integration_test_for_clickhous_exporter.yaml:10]:
```suggestion
note: Add integration test for clickhouse exporter
```

File: exporter/clickhouseexporter/README.md [hunk 2]
```diff
@@ -342,3 +342,7 @@ service:
       processors: [ batch ]
       exporters: [ clickhouse ]
 ```
+
+## develop tips
+
+- Make sure integration test is pass after any change of sql
```
Comment (codeboten) [inline exporter/clickhouseexporter/README.md:348]:
```suggestion
- Make sure integration tests pass after any change of sql
```

File: exporter/clickhouseexporter/README.md [hunk 3]
```diff
@@ -342,3 +342,7 @@ service:
       processors: [ batch ]
       exporters: [ clickhouse ]
 ```
+
+## develop tips
```
Comment (codeboten) [inline exporter/clickhouseexporter/README.md:346]:
```suggestion
## Developer tips
```

**Review by codeboten — APPROVED** [review]

**Comment by codeboten** [comment]
@Frapschen please resolve the conflicts

**Comment by Frapschen** [comment]
@codeboten ready to merge

**Comment by JaredTan95** [comment]
@Frapschen conflicts need to be solved~

**Comment by Frapschen** [comment]
unknown reason markdown-link-check job fail: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8228824833/job/22498927737?pr=30209
```
/usr/local/lib/node_modules/markdown-link-check/markdown-link-check:97
                    if (program.opts().ignore.some((ignorePath) => resolved.includes(ignorePath))) {
                                             ^

TypeError: Cannot read properties of undefined (reading 'some')
    at Command.<anonymous> (/usr/local/lib/node_modules/markdown-link-check/markdown-link-check:97:46)
    at Command.listener [as _actionHandler] (/usr/local/lib/node_modules/markdown-link-check/node_modules/commander/lib/command.js:482:17)
    at /usr/local/lib/node_modules/markdown-link-check/node_modules/commander/lib/command.js:[12](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8228824833/job/22498927737?pr=30209#step:10:13)83:65
    at Command._chainOrCall (/usr/local/lib/node_modules/markdown-link-check/node_modules/commander/lib/command.js:1[17](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8228824833/job/22498927737?pr=30209#step:10:18)7:12)
    at Command._parseCommand (/usr/local/lib/node_modules/markdown-link-check/node_modules/commander/lib/command.js:1283:27)
    at Command.parse (/usr/local/lib/node_modules/markdown-link-check/node_modules/commander/lib/command.js:909:10)
    at getInputs (/usr/local/lib/node_modules/markdown-link-check/markdown-link-check:108:7)
    at main (/usr/local/lib/node_modules/markdown-link-check/markdown-link-check:252:20)

Node.js v[18](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8228824833/job/22498927737?pr=30209#step:10:19).[19](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8228824833/job/22498927737?pr=30209#step:10:20).1
Check that anchor links are lowercase
Error: Process completed with exit code 1.
```

File: exporter/clickhouseexporter/README.md [hunk 4]
```diff
@@ -342,3 +342,7 @@ service:
       processors: [ batch ]
       exporters: [ clickhouse ]
 ```
+
+## Developer tips
+
+- Make sure integration tests pass after any change of sql
```
Comment (dmitryax) [inline exporter/clickhouseexporter/README.md:348]: I don't think it's useful without instructions on running the integration tests. The integration tests should be run by the CI

File: exporter/clickhouseexporter/exporter_logs_test.go [hunk 5]
```diff
@@ -121,6 +121,7 @@ func TestExporter_pushLogsData(t *testing.T) {
 	})
 }
 
+// nolint:unparam // not need to check this func
```
Comment (dmitryax) [inline exporter/clickhouseexporter/exporter_logs_test.go:124]: Why is this needed? All the parameters seem to be used...

File: exporter/clickhouseexporter/exporter_metrics_test.go [hunk 6]
```diff
@@ -148,6 +148,7 @@ func simpleMetrics(count int) pmetric.Metrics {
 	sm.Scope().SetDroppedAttributesCount(10)
 	sm.Scope().SetName("Scope name 1")
 	sm.Scope().SetVersion("Scope version 1")
+	timestamp := time.Unix(1703498029, 0)
```
Comment (dmitryax) [inline exporter/clickhouseexporter/exporter_metrics_test.go:151]: How is this related to the integration test?

File: .chloggen/add_integration_test_for_clickhous_exporter.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: clickhouseexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add integration test for clickhouse exporter
```
Comment (dmitryax) [inline .chloggen/add_integration_test_for_clickhous_exporter.yaml:10]: We don't need changelogs for improving test coverage. It's not a user-facing or API change

**Comment by dmitryax** [comment]
I'm merging. My comments are mostly nits. Feel free to reply later. I would still remove the changelog entry in a separate PR

**Comment by dmitryax** [comment]
Thank you for adding the test!

File: exporter/clickhouseexporter/exporter_logs_test.go (hunk 5 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_logs_test.go:124]: make lint:
image

File: exporter/clickhouseexporter/exporter_metrics_test.go (hunk 6 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_metrics_test.go:151]: integration test will verify the inserted data, so we need a constant timestamp

File: exporter/clickhouseexporter/exporter_logs_test.go (hunk 5 above)
Comment (dmitryax) [reply exporter/clickhouseexporter/exporter_logs_test.go:124]: Ok, so why do you need `fns` argument if it's never passed?
